File: app/schemas/onboarding.py
"""Onboarding checklist API schemas (PP-HRD-FO-14)."""

from __future__ import annotations

from datetime import date, datetime
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.common import BlankOptionalFieldsMixin, TimestampSchema

OnboardingItemStatus = Literal["pending", "completed", "not_applicable"]
OnboardingChecklistStatus = Literal["in_progress", "completed", "cancelled"]


class OnboardingChecklistCreate(BlankOptionalFieldsMixin, BaseModel):
    employee_name: str = Field(min_length=2, max_length=200)
    employee_user_id: Optional[UUID] = None
    employee_email: Optional[str] = Field(default=None, max_length=255)
    employee_code: Optional[str] = Field(default=None, max_length=40)
    joining_date: Optional[date] = None
    designation: Optional[str] = Field(default=None, max_length=120)
    department_name: Optional[str] = Field(default=None, max_length=120)
    org_department_id: Optional[UUID] = None
    team_id: Optional[UUID] = None
    role_id: Optional[UUID] = None
    reporting_manager_id: Optional[UUID] = None
    reporting_manager_name: Optional[str] = Field(default=None, max_length=200)
    notes: Optional[str] = None
    template_id: Optional[UUID] = None


class OnboardingChecklistUpdate(BlankOptionalFieldsMixin, BaseModel):
    employee_name: Optional[str] = Field(default=None, max_length=200)
    employee_user_id: Optional[UUID] = None
    employee_code: Optional[str] = Field(default=None, max_length=40)
    joining_date: Optional[date] = None
    designation: Optional[str] = Field(default=None, max_length=120)
    department_name: Optional[str] = Field(default=None, max_length=120)
    org_department_id: Optional[UUID] = None
    team_id: Optional[UUID] = None
    role_id: Optional[UUID] = None
    reporting_manager_id: Optional[UUID] = None
    reporting_manager_name: Optional[str] = Field(default=None, max_length=200)
    notes: Optional[str] = None
    status: Optional[OnboardingChecklistStatus] = None


class OnboardingItemStatusUpdate(BaseModel):
    status: OnboardingItemStatus
    completion_date: Optional[date] = None
    notes: Optional[str] = None


class OnboardingChecklistItemRead(TimestampSchema):
    checklist_id: UUID
    section: str
    sort_order: int
    item_text: str
    responsibility: str
    responsibility_label: str
    status: str
    status_label: str
    owner_user_id: Optional[UUID] = None
    owner_name: Optional[str] = None
    help_ticket_id: Optional[UUID] = None
    help_ticket_number: Optional[str] = None
    completed_by_id: Optional[UUID] = None
    completed_by_name: Optional[str] = None
    completion_date: Optional[date] = None
    notes: Optional[str] = None
    can_edit: bool = False
    is_mine: bool = False


class OnboardingTriggeredTicket(BaseModel):
    responsibility: str
    responsibility_label: str
    ticket_id: UUID
    ticket_number: str
    category: str
    assignee_id: Optional[UUID] = None
    item_count: int = 0


class OnboardingChecklistRead(TimestampSchema):
    template_id: Optional[UUID] = None
    template_code: Optional[str] = None
    template_name: Optional[str] = None
    employee_user_id: Optional[UUID] = None
    employee_name: str
    employee_code: Optional[str] = None
    joining_date: Optional[date] = None
    designation: Optional[str] = None
    department_name: Optional[str] = None
    org_department_id: Optional[UUID] = None
    team_id: Optional[UUID] = None
    team_name: Optional[str] = None
    role_id: Optional[UUID] = None
    role_name: Optional[str] = None
    reporting_manager_id: Optional[UUID] = None
    reporting_manager_name: Optional[str] = None
    status: str
    status_label: str
    notes: Optional[str] = None
    created_by_id: Optional[UUID] = None
    created_by_name: Optional[str] = None
    completed_at: Optional[datetime] = None
    is_published: bool = False
    published_at: Optional[datetime] = None
    published_by_id: Optional[UUID] = None
    total_items: int = 0
    completed_items: int = 0
    pending_items: int = 0
    completion_percent: int = 0
    can_manage: bool = False
    my_pending_items: int = 0


class OnboardingChecklistDetailRead(OnboardingChecklistRead):
    items: list[OnboardingChecklistItemRead] = Field(default_factory=list)
    sections: list[str] = Field(default_factory=list)
    triggered_tickets: list[OnboardingTriggeredTicket] = Field(default_factory=list)
    # Set only when onboarding auto-created the User (not when linking an existing account).
    provisioned_temporary_password: Optional[str] = None


class OnboardingTemplateRead(BaseModel):
    id: UUID
    code: str
    name: str
    version: int
    is_active: bool

File: frontend/src/api/onboarding.ts
import { apiClient } from './client';

export type OnboardingItemStatus = 'pending' | 'completed' | 'not_applicable';
export type OnboardingChecklistStatus = 'in_progress' | 'completed' | 'cancelled';

export interface OnboardingChecklistItem {
  id: string;
  checklist_id: string;
  section: string;
  sort_order: number;
  item_text: string;
  responsibility: string;
  responsibility_label: string;
  status: OnboardingItemStatus;
  status_label: string;
  owner_user_id: string | null;
  owner_name: string | null;
  help_ticket_id: string | null;
  help_ticket_number: string | null;
  completed_by_id: string | null;
  completed_by_name: string | null;
  completion_date: string | null;
  notes: string | null;
  can_edit: boolean;
  is_mine: boolean;
  created_at: string;
  updated_at: string;
}

export interface OnboardingTriggeredTicket {
  responsibility: string;
  responsibility_label: string;
  ticket_id: string;
  ticket_number: string;
  category: string;
  assignee_id: string | null;
  item_count: number;
}

export interface OnboardingChecklist {
  id: string;
  template_id: string | null;
  template_code: string | null;
  template_name: string | null;
  employee_user_id: string | null;
  employee_name: string;
  employee_code: string | null;
  joining_date: string | null;
  designation: string | null;
  department_name: string | null;
  org_department_id: string | null;
  team_id: string | null;
  team_name: string | null;
  role_id: string | null;
  role_name: string | null;
  reporting_manager_id: string | null;
  reporting_manager_name: string | null;
  status: OnboardingChecklistStatus;
  status_label: string;
  notes: string | null;
  created_by_id: string | null;
  created_by_name: string | null;
  completed_at: string | null;
  is_published: boolean;
  published_at: string | null;
  published_by_id: string | null;
  total_items: number;
  completed_items: number;
  pending_items: number;
  completion_percent: number;
  can_manage: boolean;
  my_pending_items: number;
  created_at: string;
  updated_at: string;
}

export interface OnboardingChecklistDetail extends OnboardingChecklist {
  items: OnboardingChecklistItem[];
  sections: string[];
  triggered_tickets: OnboardingTriggeredTicket[];
}

export interface OnboardingChecklistCreate {
  employee_name: string;
  employee_user_id?: string | null;
  employee_email?: string | null;
  employee_code?: string | null;
  joining_date?: string | null;
  designation?: string | null;
  department_name?: string | null;
  org_department_id?: string | null;
  team_id?: string | null;
  role_id?: string | null;
  reporting_manager_id?: string | null;
  reporting_manager_name?: string | null;
  notes?: string | null;
  template_id?: string | null;
}

export interface OnboardingChecklistUpdate {
  employee_name?: string;
  employee_user_id?: string | null;
  employee_code?: string | null;
  joining_date?: string | null;
  designation?: string | null;
  department_name?: string | null;
  org_department_id?: string | null;
  team_id?: string | null;
  role_id?: string | null;
  reporting_manager_id?: string | null;
  reporting_manager_name?: string | null;
  notes?: string | null;
  status?: OnboardingChecklistStatus;
}

function buildQuery(params?: Record<string, string | undefined>) {
  if (!params) return '';
  const qs = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    if (value != null && value !== '') qs.set(key, value);
  });
  const raw = qs.toString();
  return raw ? `?${raw}` : '';
}

export const onboardingApi = {
  list: async (params?: { status?: string }): Promise<OnboardingChecklist[]> => {
    const { data } = await apiClient.get<OnboardingChecklist[]>(
      `/hr/onboarding${buildQuery(params)}`,
    );
    return data;
  },
  get: async (id: string): Promise<OnboardingChecklistDetail> => {
    const { data } = await apiClient.get<OnboardingChecklistDetail>(`/hr/onboarding/${id}`);
    return data;
  },
  create: async (payload: OnboardingChecklistCreate): Promise<OnboardingChecklistDetail> => {
    const { data } = await apiClient.post<OnboardingChecklistDetail>('/hr/onboarding', payload);
    return data;
  },
  update: async (
    id: string,
    payload: OnboardingChecklistUpdate,
  ): Promise<OnboardingChecklistDetail> => {
    const { data } = await apiClient.patch<OnboardingChecklistDetail>(
      `/hr/onboarding/${id}`,
      payload,
    );
    return data;
  },
  delete: async (id: string): Promise<void> => {
    await apiClient.delete(`/hr/onboarding/${id}`);
  },
  publish: async (id: string): Promise<OnboardingChecklistDetail> => {
    const { data } = await apiClient.post<OnboardingChecklistDetail>(
      `/hr/onboarding/${id}/publish`,
    );
    return data;
  },
  setItemStatus: async (
    checklistId: string,
    itemId: string,
    payload: {
      status: OnboardingItemStatus;
      completion_date?: string | null;
      notes?: string | null;
    },
  ): Promise<OnboardingChecklistDetail> => {
    const { data } = await apiClient.post<OnboardingChecklistDetail>(
      `/hr/onboarding/${checklistId}/items/${itemId}/status`,
      payload,
    );
    return data;
  },
};
